src/common/Cryptography/HMAC.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of the TrinityCore Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TRINITY_HMAC_H
+#define TRINITY_HMAC_H
+
+#include "CryptoConstants.h"
+#include "Define.h"
+#include "Errors.h"
+#include <array>
+#include <string>
+#include <string_view>
+#include <openssl/hmac.h>
+
+class BigNumber;
+
+namespace Trinity::Impl
+{
+    struct HMACImpl
+    {
+        typedef EVP_MD const* (*HashCreator)();
+
+#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER < 0x10100000L
+        static HMAC_CTX* MakeCTX()
+        {
+            HMAC_CTX* ctx = new HMAC_CTX();
+            HMAC_CTX_init(ctx);
+            return ctx;
+        }
+
+        static void DestroyCTX(HMAC_CTX* ctx)
+        {
+            HMAC_CTX_cleanup(ctx);
+            delete ctx;
+        }
+#else
+        static HMAC_CTX* MakeCTX() { return HMAC_CTX_new(); }
+        static void DestroyCTX(HMAC_CTX* ctx) { HMAC_CTX_free(ctx); }
+#endif
+    };
+
+    template <HMACImpl::HashCreator HashCreator, size_t DigestLength>
+    class GenericHMAC
+    {
+        public:
+            static constexpr size_t DIGEST_LENGTH = DigestLength;
+            using Digest = std::array<uint8, DIGEST_LENGTH>;
+
+            template <typename C>
+            static Digest GetDigestOf(C const& seed, uint8 const* data, size_t len)
+            {
+                GenericHMAC hash(seed);
+                hash.UpdateData(data, len);
+                hash.Finalize();
+                return hash.GetDigest();
+            }
+
+            template <typename C, typename... Ts>
+            static auto GetDigestOf(C const& seed, Ts&&... pack) -> std::enable_if_t<!(std::is_integral_v<std::decay_t<Ts>> || ...), Digest>
+            {
+                GenericHMAC hash(seed);
+                (hash.UpdateData(std::forward<Ts>(pack)), ...);
+                hash.Finalize();
+                return hash.GetDigest();
+            }
+
+            GenericHMAC(uint8 const* seed, size_t len) : _ctx(HMACImpl::MakeCTX())
+            {
+                int result = HMAC_Init_ex(_ctx, seed, len, HashCreator(), nullptr);
+                ASSERT(result == 1);
+            }
+            template <typename C> GenericHMAC(C const& container) : GenericHMAC(std::data(container), std::size(container)) {}
```

Comment:
Why are you placing the template part on the same line as everything else? We don't do that anywhere